plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,239 @@
+				Address: net.JoinHostPort(u.Hostname(), port),
+			}
+		}
+
+	case "udp":
+		// Use default UDP port if it was not specified
+		port := u.Port()
+		if port == "" {
+			port = "14550"
+		}
+
+		if u.Hostname() == "" {
+			m.endpointConfig = gomavlib.EndpointUDPServer{
+				Address: "0.0.0.0:" + port,
+			}
+		} else {
+			m.endpointConfig = gomavlib.EndpointUDPClient{
+				Address: net.JoinHostPort(u.Hostname(), port),
+			}
+		}
+
+	default:
+		return fmt.Errorf("unknown scheme %q", u.Scheme)
+	}
+
+	// Compile filter
+	m.filter, err = filter.Compile(m.FilterPattern)
+	if err != nil {
+		return fmt.Errorf("compiling filter failed: %w", err)
+	}
+
+	return nil
+}
+
+func (m *Mavlink) Start(acc telegraf.Accumulator) error {
+	// Start MAVLink endpoint
+	connection, err := gomavlib.NewNode(gomavlib.NodeConf{
+		Endpoints:              []gomavlib.EndpointConf{m.endpointConfig},
+		Dialect:                ardupilotmega.Dialect,
+		OutVersion:             gomavlib.V2,
+		OutSystemID:            m.SystemID,
+		StreamRequestEnable:    m.StreamRequestFrequency > 0,
+		StreamRequestFrequency: int(m.StreamRequestFrequency),
+	})
+	if err != nil {
+		return &internal.StartupError{
+			Err:   fmt.Errorf("connecting to mavlink endpoint %s failed: %w", m.URL, err),
+			Retry: true,
+		}
+	}
+	m.connection = connection
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	m.cancel = cancelFunc
+
+	// Start routine to connect to Mavlink and stream out data async
+	m.wg.Add(1)
+	go func(ctx context.Context) {
+		defer m.connection.Close()
+		defer m.wg.Done()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case evt := <-m.connection.Events():
+				switch evt := evt.(type) {
+				case *gomavlib.EventFrame:
+					if result := convertFrameToMetric(evt.Frame, m.filter); result != nil {
+						result.AddTag("source", m.URL)
+						acc.AddMetric(result)
+					}
+				case *gomavlib.EventChannelOpen:
+					m.Log.Tracef("Mavlink channel opened")
+				case *gomavlib.EventChannelClose:
+					m.Log.Tracef("Mavlink channel closed")
+				case *gomavlib.EventParseError:
+					m.Log.Tracef("Mavlink parse error: %s", evt.Error.Error())
+				case *gomavlib.EventStreamRequested:
+					m.Log.Tracef("Issued stream request to system %d, component %d", evt.SystemID, evt.ComponentID)
+				default:
+					m.Log.Tracef("Unhandled Mavlink event: %T", evt)
+				}
+			}
+		}
+	}(ctx)
+
+	return nil
+}
+
+func (*Mavlink) Gather(telegraf.Accumulator) error {
+	return nil
+}
+
+func (m *Mavlink) Stop() {
+	m.cancel()
+	m.wg.Wait()
+}
+
+// Convert a Mavlink frame into a struct containing Metric data.
+func convertFrameToMetric(frm frame.Frame, msgFilter filter.Filter) telegraf.Metric {
+	m := frm.GetMessage()
+	t := reflect.TypeOf(m)
+	v := reflect.ValueOf(m)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		v = v.Elem()
+	}
+
+	name := internal.SnakeCase(strings.TrimPrefix(t.Name(), "Message"))
+
+	if msgFilter != nil && !msgFilter.Match(name) {
+		return nil
+	}
```

Comment:
With the suggestion above this converts to
```suggestion
	if m.filter != nil && !m.filter.Match(name) {
		return
	}
```